From: Shyamanth RH <shyamanth.rh@arm.com>
Date: Mon, 17 Aug 2026 00:00:00 +0000
Subject: [PATCH] rdv3: integrate RME ACS EL3 support

Extend the RDV3 address space and reserve the first 2 MiB of DRAM2 as Root
memory for ACS. The ACS EL3 implementation is linked separately after the
normal TF-A build.

Signed-off-by: Shyamanth RH <shyamanth.rh@arm.com>
---
 plat/arm/board/neoverse_rd/common/include/nrd3/nrd_pas_def3.h | 2 +-
 .../neoverse_rd/common/include/nrd3/nrd_plat_arm_def3.h        | 14 ++++++++++----
 plat/arm/board/neoverse_rd/common/nrd_plat3.c                  | 1 +
 plat/arm/board/neoverse_rd/platform/rdv3/include/platform_def.h| 2 +-
 4 files changed, 13 insertions(+), 6 deletions(-)
diff --git a/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_pas_def3.h b/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_pas_def3.h
--- a/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_pas_def3.h
+++ b/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_pas_def3.h
@@ -572,7 +572,7 @@
 		GPT_MAP_REGION_GRANULE(					\
 			NRD_MC_BASE(NRD_CSS_DRAM2_BASE, 0),		\
 			ARM_DRAM2_SIZE,					\
-			GPT_GPI_NS)
+			GPT_GPI_ANY)
 
 #define NRD_PAS_DRAM2_CHIP1						\
 		GPT_MAP_REGION_GRANULE(					\
diff --git a/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_plat_arm_def3.h b/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_plat_arm_def3.h
--- a/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_plat_arm_def3.h
+++ b/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_plat_arm_def3.h
@@ -56,8 +56,8 @@
  * chips are accessed - secure ram, css device and soc device regions.
  */
 #if defined(IMAGE_BL31)
-#  define PLAT_ARM_MMAP_ENTRIES		(13 + ((NRD_CHIP_COUNT - 1) * 3))
-#  define MAX_XLAT_TABLES		(13 + ((NRD_CHIP_COUNT - 1) * 3))
+#  define PLAT_ARM_MMAP_ENTRIES		(14 + ((NRD_CHIP_COUNT - 1) * 3))
+#  define MAX_XLAT_TABLES		(14 + ((NRD_CHIP_COUNT - 1) * 3))
 #elif defined(IMAGE_BL32)
 # define PLAT_ARM_MMAP_ENTRIES		U(8)
 # define MAX_XLAT_TABLES		U(5)
@@ -891,8 +891,14 @@ typedef enum sds_region_id {
 
 #define ARM_MAP_DRAM2							\
 		MAP_REGION_FLAT(					\
-			ARM_DRAM2_BASE,					\
-			ARM_DRAM2_SIZE,					\
+			ARM_DRAM2_BASE + SZ_2M,				\
+			ARM_DRAM2_SIZE - SZ_2M,				\
 			MT_MEMORY | MT_RW | MT_NS)
 
+#define ARM_MAP_DRAM2_ROOT						\
+		MAP_REGION_FLAT(					\
+			ARM_DRAM2_BASE,					\
+			SZ_2M,						\
+			MT_MEMORY | MT_RW | MT_ROOT)
+
 #endif /* NRD_PLAT_ARM_DEF3_H */
diff --git a/plat/arm/board/neoverse_rd/common/nrd_plat3.c b/plat/arm/board/neoverse_rd/common/nrd_plat3.c
--- a/plat/arm/board/neoverse_rd/common/nrd_plat3.c
+++ b/plat/arm/board/neoverse_rd/common/nrd_plat3.c
@@ -73,6 +73,7 @@ const mmap_region_t plat_arm_mmap[] = {
 #if IMAGE_BL31
 const mmap_region_t plat_arm_mmap[] = {
 	NRD_CSS_SHARED_RAM_MMAP(0),
+	ARM_MAP_DRAM2_ROOT,
 #if NRD_ERR_INJ_SUPPORT
 	ARM_MAP_EINJ_DRAM1,
 #endif
diff --git a/plat/arm/board/neoverse_rd/platform/rdv3/include/platform_def.h b/plat/arm/board/neoverse_rd/platform/rdv3/include/platform_def.h
--- a/plat/arm/board/neoverse_rd/platform/rdv3/include/platform_def.h
+++ b/plat/arm/board/neoverse_rd/platform/rdv3/include/platform_def.h
@@ -38,7 +38,7 @@
 #define NRD_CSS_DRAM2_SIZE		ULL(0x180000000)
 
 /* Address bits */
-#define NRD_ADDR_BITS_PER_CHIP		U(36)  /* 64GB */
+#define NRD_ADDR_BITS_PER_CHIP		U(40)  /* 1TB */
 
 /*
  * In the current implementation, the RoT Service request that requires the
-- 
2.43.0
